ch7.2.2.2-ex121/build-clause-table.ts

```diff
@@ -0,0 +1,167 @@
+interface Literal {
+    n: number;
+    isPositive: boolean;
+}
+
+function variable(n: number): [Literal, Literal] {
+    return [
+        { n, isPositive: true },
+        { n, isPositive: false },
+    ];
+}
+
+function variables(n: number): Array<[Literal, Literal]> {
+    let result: Array<[Literal, Literal]> = [];
+    for (let i = 1; i <= n; i++) {
+        result.push(variable(i));
+    }
+    return result;
+}
+
+let [[p1, n1], [p2, n2], [p3, n3], [p4, n4]] = variables(4);
```

Comment:
I found it inelegant that you need to explicitly declare the "4" in the call. Then I realized that you can turn `variables` into a generator function. (!) I tried it in the browser, and it works!

```
function* foo() { let n = 1; while (true) { yield [n, n + 1]; n += 2; } }
let [[p1, n1], [p2, n2], [p3, n3]] = foo();
p1; // 1
p3; // 5
```